Python spotcrime

jcconnell opened the pull request:
## Description:
Added support for the Spot Crime API.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4685

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: spotcrime
     name: Spot Crime
     days: 8
     radius: 0.05 # 5 miles
```

## Checklist:
  - [x] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

jcconnell committed:
Added python-spotcrime sensor

jcconnell committed:
Added python-spotcrime sensor

homeassistant commented:
Hi @jcconnell, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#12460).

Once you do that we will be able to review and accept this pull request.

Thanks!

happyleavesaoc commented:
Neat! Is this ultimately the same data as [`crimereports`](https://home-assistant.io/components/sensor.crimereports/), or is there some other advantage?

jcconnell commented:
Hi! I used your code as a template since this is my first time contributing (both to HASS and OSS).

I used crimereports in my HA installation for a while, but never received any notice of an incident. After digging in a bit further, I realized that Crime Reports doesn't receive any notifications in my area, but Spot Crime does. I don't know of their specific differences across the states but in my area, Spot Crime provides some results where Crime Reports didn't.

happyleavesaoc reviewed (changes requested):
You'll want to remove `man/man1/nosetests.1` from the PR. I think it got accidently included.

jcconnell commented:
@happyleavesaoc Thank you for the help! I'm not very familiar with this process. What's the best way to remove the file from the PR?

balloob commented:
```
git rm man/man1/nosetests.1
git commit -m "Remove accidental included file"
git push
```

jcconnell committed:
Remove accidental included file

jcconnell committed:
Update indentation to match PEP8

pvizeli commented on homeassistant/const.py:
```diff
@@ -51,6 +51,7 @@
 CONF_CUSTOMIZE = 'customize'
 CONF_CUSTOMIZE_DOMAIN = 'customize_domain'
 CONF_CUSTOMIZE_GLOB = 'customize_glob'
+CONF_DAYS = 'days'
```
Do that only if you can replace it and use it on other components/platform

jcconnell replied:
What should I remove or how should I change it to be correct?

bachya replied:
I think @pvizeli is saying that `const.py` is for constants that are broadly applicable; as far as we know, `CONF_DAYS` is only useful to your platform at the moment. So, until it's determined otherwise, it's best placed within `spotcrime.py`.

jcconnell replied:
Thank you @bachya, I see now. I did not understand well enough how to bring in the configuration variables. Some of the information at this link gave me a better idea: https://community.home-assistant.io/t/help-using-config-get-for-custom-platform-switch/4092/2. I am working to do it the correct way now.

jcconnell committed:
Remove days from const.py

jcconnell committed:
Changed default days to 1

MartinHjelmare commented on homeassistant/components/sensor/spotcrime.py:
```diff
@@ -0,0 +1,127 @@
+        import spotcrime
+        incident_counts = defaultdict(int)
+        incidents = self._spotcrime.get_incidents()
+        fire_events = len(self._previous_incidents) > 0
```
You don't need to define a new variable here. Just check if `self._previous_incidents` is truthy below.

MartinHjelmare commented on homeassistant/components/sensor/spotcrime.py:
```diff
@@ -0,0 +1,127 @@
+        for incident in incidents:
+            incident_type = slugify(incident.get('type'))
+            incident_counts[incident_type] += 1
+            if (fire_events and incident.get('id')
```
```py
if (self._previous_incidents and incident.get('id')
        not in self._previous_incidents):
```

MartinHjelmare commented on homeassistant/components/sensor/spotcrime.py:
```diff
@@ -0,0 +1,127 @@
+                self._incident_event(incident)
+            self._previous_incidents.add(incident.get('id'))
+        self._attributes = {
+            ATTR_ATTRIBUTION: spotcrime.ATTRIBUTION
```
Do this in init instead.

jcconnell committed:
Incorporate changes requested by MartinHjelmare

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Fix the lint issue and we can merge.

jcconnell commented:
@MartinHjelmare Thank you for the review! Really appreciate the detail. Removed the un-needed import and I _think_ I fixed the lint issue. Testing now.

jcconnell committed:
Remove unnecessary import and fix lint error

jcconnell commented:
Nice! Thanks for all the help on my first go everyone!
